### PR TITLE
fix: handle expired not authorized status

### PR DESCRIPTION
**Pull request by pietro-tota**
#### List of Changes
Add logic to distinguish status for the status on which update the expired transaction.
Before this modification all transaction was updated to `EXPIRED` status.
This modification check for each transaction if authorization was requested and then set expiration status to EXPIRED or EXPIRED_NOT_AUTHORIZED according to the fact that the authorization was requested or not

#### Motivation and Context
This modification is needed in order to update transaction status correctly to EXPIRED or EXPIRED_NOT_AUTHORIZED

#### How Has This Been Tested?
Junit test + test with ecommerce local

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/TransactionExpiredEventPublisher.kt [hunk 1]
```diff
@@ -31,7 +33,28 @@ class TransactionExpiredEventPublisher(
...
-    ) = publishAllEvents(baseTransactions, TransactionStatusDto.EXPIRED, batchExecutionTimeWindow)
+    ): Mono<Boolean> {
+        // split expired transaction in two lists: one for transactions without requested
+        // authorization and one with requested authorization
+        val splittedTransactions: Map<Boolean, List<BaseTransaction>> =
+            baseTransactions
+                .stream()
+                .collect(Collectors.groupingBy { it is BaseTransactionWithRequestedAuthorization })
+        val baseTransactionsWithRequestedAuthorization = splittedTransactions[true] ?: emptyList()
+        val baseTransactionsActivatedOnly = splittedTransactions[false] ?: emptyList()
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/TransactionExpiredEventPublisher.kt:44]: I don't like `splittedTransaction[true]` too much :)
What about:

```suggestion
        val (baseTransactionsWithRequestedAuthorization,  baseTransactionsActivatedOnly) = baseTransactions
                .partition { it is BaseTransactionWithRequestedAuthorization })
```

Using kotlin's [partition](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/partition.html)

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/TransactionExpiredEventPublisher.kt [hunk 2]
```diff
@@ -31,7 +33,28 @@ class TransactionExpiredEventPublisher(
...
+        val mergedTransactions =
+            baseTransactionsWithRequestedAuthorization
+                .map { Pair(it, TransactionStatusDto.EXPIRED) }
+                .plus(
```
Comment (giovanniberti) [inline src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/TransactionExpiredEventPublisher.kt:48]: If you prefer you can replace `plus()` with `+`

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/EventPublisher.kt [hunk 3]
```diff
@@ -58,19 +58,18 @@ abstract class EventPublisher<E>(
...
+        transactions: List<Pair<BaseTransaction, TransactionStatusDto>>,
         batchExecutionWindowMillis: Long
     ): Mono<Boolean> {
         val eventOffset = AtomicLong(0L)
         val offsetIncrement = batchExecutionWindowMillis / transactions.size
         return Flux.fromIterable(transactions)
             .parallel(parallelEventsToProcess)
             .runOn(Schedulers.parallel())
-            .flatMap { publishEvent(it, newStatus, eventOffset.addAndGet(offsetIncrement)) }
+            .flatMap { publishEvent(it.first, it.second, eventOffset.addAndGet(offsetIncrement)) }
```
[the remark this answers is not in the document]
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/EventPublisher.kt]: Fixed with https://github.com/pagopa/pagopa-ecommerce-transactions-scheduler-service/pull/19/commits/514c79875ac2b6951e0ac028cd2c19a0ba2fa923

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/TransactionExpiredEventPublisher.kt (hunk 2 above)
Comment (pietro-tota) [reply src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/publishers/TransactionExpiredEventPublisher.kt:43]: Fixed with https://github.com/pagopa/pagopa-ecommerce-transactions-scheduler-service/pull/19/commits/514c79875ac2b6951e0ac028cd2c19a0ba2fa923

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**